### PR TITLE
Bug 1845905 - Add a utility to create store lazily

### PR DESCRIPTION
**Pull request by rahulsainani**
### What
Similar to how `by viewModels()` work, this adds a utility to create stores in `ViewModelStoreOwner`s like `Fragment` or `Activity`.

### Why
– To not use late init var to have the property in a bad state early on with a promise that it will be created later.
– To make sure that `StoreProvider` is used and the `Store` will survive the configuration change.
– To prevent accidental use of `LazyThreadSafetyMode.SYNCHRONIZED` by default.

`lazy` uses `LazyThreadSafetyMode.SYNCHRONIZED` thread safety mode by default. While that may be fine for singleton (app scoped) stores, but for fragment scoped stores, as we're sure that the initialization (and access) will always happen in the same (main) thread, as the one where you use the property, we can use `LazyThreadSafetyMode.NONE`. It doesn't incur any thread-safety guarantees and related overhead.

### Usage 
From:
```
private val store by lazy(LazyThreadSafetyMode.NONE) {
    StoreProvider.get(this) {
        MyStore()
    }
}
```

The syntax gets simplified to:

```
val myStore by store { MyStore() }
```

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1845905

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/components/StoreProvider.kt [hunk 1]
```diff
@@ -42,3 +45,37 @@ class StoreProviderFactory<T : Store<*, *>>(
         return StoreProvider(createStore()) as VM
     }
 }
+
+/**
+ * Helper function for creation of [StoreLazy].
+ */
+@MainThread
+fun <T : Store<*, *>> ViewModelStoreOwner.store(
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/components/StoreProvider.kt:53]: thoughts?
```suggestion
fun <T : Store<*, *>> ViewModelStoreOwner.lazyStore(
```

**Review by MozillaNoah — APPROVED** [review]
this looks great! just left one question about the naming

File: fenix/app/src/main/java/org/mozilla/fenix/components/StoreProvider.kt (hunk 1 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/components/StoreProvider.kt]: Yeah makes sense, update to this and simplified the implementation.

**Comment by jonalmeida** [comment]
@rahulsainani @MozillaNoah hey folks this is a foundational change with how we will use the store that flew under the radar:

1. Let's not land a change like this without getting more eyes on it. Please feel free to cc other team members too that have worked extensively on the code base at this level.
2. We've landed utility code and not any usage of it. The closest I've seen is [this PR](https://github.com/mozilla-mobile/firefox-android/pull/3153/) but no usage of it. Adding things that we don't have clear understanding how they will be used is not great.

**Comment by rahulsainani** [comment]
As discussed on the call about the need & utility for these changes, I'll create the parent bug with the story about these 2 changes. Appreciate your perspective. We can surely try to get more eyes on patches like this in the future.

**Pull request merged**